xwiki-commons-tools/xwiki-commons-tool-xar/xwiki-commons-tool-xar-integration-tests/src/test/resources/transformedXml/src/main/resources/Blog/BlogClass.xwikiclass

```diff
@@ -0,0 +1,99 @@
+<class>
```

Comment:
I would name the file *.xml since it contains XML.

Also would be good to add a license header (we need to add that to other files too).

Reply:
Ok for the license header.
These files end-up into the XAR... (it could or should be avoided) thus they become pages if they are named `.xml` which would fail since there is no `xwikidoc` element. Ok to keep `.xwikiclass`?
(there's a lot more XML files out there which are not called `.xml`, e.g. the `.plist` on macs, or the rss files).

Reply:
I don't remember the test but then it looks like these resource files are placed in the wrong location if they end up in the XAR when they shouldn't. Using a different suffix is not the right way IMO. In XWiki we use xml for xml files (it helps in the IDE, it helps visually know what the file is about, etc). I don't see any valid argument for not doing that honestly, especially since this file is a xwiki XML page which is *supposed* to end in `.xml`.

Reply:
I've now checked the code and the doc at https://dev.xwiki.org/xwiki/bin/view/Community/XARPlugin#HTransformations. I think the doc is a bit wrong at https://dev.xwiki.org/xwiki/bin/view/Community/XARPlugin#HTransformations and the files that are used only as intermediary resources that are used for transformations should be excluded. If they're not needed to be processed (velocity doesn't need to run on them) they could also be put in a different directory such as `src/main/xar`. I think that's a better idea than the exclude.

Reply:
> they could also be put in a different directory such as src/main/xar. I think that's a better idea than the exclude.

Yes, I don't think exclusion is a good idea.

Reply:
License header fixed. It was also missing in WebHome.xml.

Reply:
> In XWiki we use xml for xml files

That's not true for svg files.

Reply:
svg is a well-know format and extension. That's not an issue.

Reply:
> License header fixed. It was also missing in WebHome.xml.

Yes we still need to fix more places and activate the check for them since we're legally bound to have license headers in all non-binary files AFAIU.

Reply:
>>>   In XWiki we use xml for xml files
>> That's not true for svg files.
> svg is a well-know format and extension. That's not an issue.

No-one said it's an issue. 
My issue is that you consider that all files that are XML in nature are named with the extension `.xml`.

Reply:
You're only taking 1/3 of my arguments... My point was that your naming is not great and using that as a workaround of a problem is even less good (especially when solving it is very easy).

Reply:
To me the best is to use an expressive extension to describe the content.
`.xwikiclass` is expressive to me. `.xml` (or `.txt` for structured data, or `.json`...) is not. We obviously disagree on this.

The fact that using `.xml` is a blocker is more a problem of the XARMojo, I believe. If we can avoid that, this is good. If we cannot, considering a split is a good idea but may speak in favour of using a different name.

I described in some other place that I believe it is best practice to have these elements within the same directory and to fix the XARMojo so that the files are not copied. Are you disputing this?

Reply:
> To me the best is to use an expressive extension to describe the content.
.xwikiclass is expressive to me. .xml (or .txt for structured data, or .json...) is not. We obviously disagree on this.

Feel free to make a proposal on the forum so that we replace all our .xml extensions for all the XWiki XML pages we store everywhere. I doubt it would be accepted though.

Reply:
UPDATE: I just noticed that your XML file is not a full XWiki XML file format (I thought it was). So I'm tuning down my comment. I still believe it's nicer to have an xml suffix but I'm fine to accept another name. However we still need to improve it so that these files don't end up in target/classes or in the XAR, independently of their names.

Reply:
Very cool. Btw, the file name is entirely left to the developer. If you give the XARMojo a JPEG file as XML source for a REPLACE or INSERT_CHILD, it will (try to) parse it.

So we should open a new bug about the XARMojo so that it implicitly adds excludes for files that are not considered as pages, right? (we could even try to parse the document if xml and exclude it if is not with root `xwikidoc` once we find this method).

Reply:
> So we should open a new bug about the XARMojo so that it implicitly adds excludes for files that are not considered as pages, right? (we could even try to parse the document if xml and exclude it if is not with root xwikidoc once we find this method).

This is not the problem and there's no bug to open on XARMojo. It already only reads xml files. The problem is the usage of the XARMojo (ie the tests you're modifying). They must not mix files that are sources for the XAR from files that are sources for the transformations. It's just bad practice and asking for trouble to mix them. It's easy to see that. Imagine if you had named your file .xml, you'd see the issue immediately.

What is the part that you don't understand or don't agree with? I'm asking because we're turning in circles and I feel like I keep repeating myself over and over.

Reply:
1.- this discussion should not be here...
2.- I don't agree with the fact that XARMojo should copy vm, html, png, properties or any other file since it does nothing with this
3.- I don't agree with your view that it is best practice to separate the folders between what's the input for the transformation (e.g. properties or vm) and what is not quite the input (e.g. the XML which is, in the absence of transformation, just copied).

Reply:
Ok so let me redo all the analysis work from scratch (I didn't want to spend so much time on this since it was minor and simple for me):

1) the source for the XML files is `target/classes` (https://github.com/xwiki/xwiki-commons/blob/348e478f27b18e9a2894770d7dc46bac145a8707/xwiki-commons-tools/xwiki-commons-tool-xar/xwiki-commons-tool-xar-plugin/src/main/java/org/xwiki/tool/xar/XARMojo.java#L111). And not `src/main/resources` (except for `package.xml` but that's potentially an error IMO).
2) The transformation sources are taken from `src/main/resources` in the existing code and in your PR:
   ```
            <transformation>
              <file>Blog/WebHome.xml</file>
              <xpath>/xwikidoc/content</xpath>
              <action>INSERT_TEXT</action>
              <charset>UTF-8</charset>
              <content>src/main/resources/Blog/WebHome.vm</content>
            </transformation>
   ```

Conclusions:
1) There's a problem since the maven resources plugin will copy all files from `src/main/resources` to `target/classes` while the transformation sources are taken from `src/main/resources` which is useless in this case (and prevents using filtering if needed BTW) 
2) There's another problem since the maven resources plugin will copy all files from `src/main/resources` to `target/classes` including XML files named `*.xml`, which will then be included in the XAR by default (https://github.com/xwiki/xwiki-commons/blob/71bbc482c5901baa61902511be5b6ed84b29b902/xwiki-commons-tools/xwiki-commons-tool-xar/xwiki-commons-tool-xar-plugin/src/main/java/org/xwiki/tool/xar/AbstractXARMojo.java#L93).

Possible solutions:
1) Configure the maven resource plugin to exclude transformation files from being copied to `target/classes`. Note that this works unless we want to use the filtering feature of the maven resources plugin (not the case in this test).
2) Use a separate source directories for transformation files (`src/main/transformations`). This will make sure that the maven resource plugin will not copy them and that the XARMojo won't try to include them in the XAR.
3) Change the `pom.xml` of the test to not use `src/main/resources` for transformation content and instead fetch them from `target/classes` (since they've been copied, and they can benefit from filtering too). Configure the XARMojo plugin with includes/excludes to only include what's needed in the generated XAR.

Pros/cons:
* Solution 1 is not very interesting IMO. No filtering and a bit complex since you need to hand pick the excludes.
* Solution 2 is the best to me (unless we want filtering but right now we've never used that so it's pretty safe and even if needed if doesn't prevent using solution 1 or 3 in the rare cases where it'd be needed), because it's very simple and doesn't require dealing with hand-picked includes/excludes. So it's simpler and less prone to errors. 
* Solution 3 is good if we need filtering. But same as solution 1, it requires overriding the include/excludes and hand-pick them, which is more complex and error prone.

So I continue to think that solution 2 is the best (@tmortagne seems to think too according to https://github.com/xwiki/xwiki-commons/pull/136#discussion_r586416748).

Now you said you don't like it but I don't recall seing any argument from you as to why? Which solution do you prefer and why?

Thanks

Reply:
Hello @vmassol ,

I don't like your solution 2 because files that constitute other files should be aside of them. This makes sense, for example, when vm should include neighbour pages... It is disputable and the dispute is really a matter of "conceptual cleanliness vs tools current capabilities". That said, I must say that if you insist that copied files inside a xar are a problem, I guess it is indeed (as @tmortagne mentioned) the lesser evil.

I mentioned my wish for a coherent picture in [this more generic post](https://forum.xwiki.org/t/xar-mojo-best-practice-for-organizing-a-xar-project/8380). And I really think we should not argue further on this pull-request as it was just trying to fix a bug in less than two lines while the debate is far broader and touches much older things (e.g. the transformation examples or best practices more than the tests).

paul